<!-- RECURSIVE LAUNCH FILE TO INITIALIZE ALL THE NODES IN THE CORRECT NAMESPACE -->

<!-- TODO: check if the recursion is correct -->
<launch>

    <!-- ARGUMENT VALUE HOLDING THE TOTAL NUMBER OF ROBOTS, IT WILL BE PASSED FROM THE PYTHON SCRIPT -->
    <arg name="num_unicycles" default="1" />
    <arg name="num_drones" default="1" />

    <!-- UNICYCLE NODES -->

    <group ns="unicycles">
        <arg name="ns" default="$(arg num_unicycles)" />
        <!-- I CREATE THE FIRST NAMESPACE STARTING FROM THE LAST ROBOT -->
        <group ns="$(eval 'unicycle' + str(arg('ns')))">
        <!-- HERE WILL BE STARTED ALL THE NECESSARY NODES FOR EACH ROBOT -->
        
            <!-- NODE THAT SIMULATES THE UWB TAGS'S DISTANCE FROM EACH ROBOT -->
            <node name="uwb_dist_sim" pkg="dist_project" type="uwb_dist_sim.py" />

            <!-- NODE THAT HANDLES THE MOTION PLANNING OF THE ROBOT -->
            <node name="motion_planner" pkg="dist_project" type="motion_planner.py" />

            <!-- NODE THAT HANDLES THE EGO-LOCALIZATION VIA A KALMAN FILTER -->
            <node name="kalman_localization" pkg="dist_project" type="kalman_localization.py" />

            <!-- NODE THAT HANDLES THE IMAGE PROCESSING OF THE RAW IMAGE DATA COMING FROM GAZEBO SIMULATION -->
            <node name="image_processing" pkg="dist_project" type="image_processing.py" />

            <!-- NODE THAT HANDLES THE TARGET ESTIMATION -->
            <node name="target_estimator" pkg="dist_project" type="target_estimator.py" >
                <param name="namespace" value="$(arg ns)" />
            </node>
        
        <!-- NODE THAT IF UNCOMMENTED LAUNCH A SIMPLE TERMINAL CONTROLLER FOR THE ROBOT, USED FOR PLOTTING PURPOSES -->
        <!-- <node name="teleop_twist_keyboard" pkg="teleop_twist_keyboard" type="teleop_twist_keyboard.py" /> -->
        </group>
        <include file="$(find dist_project)/launch/start_nodes.launch" if="$(eval arg('ns') - 1 >= 0)">
            <arg name="num_unicycles" value="$(eval arg('ns') - 1)" />
            <arg name="num_drones" value="$(arg num_drones)" />
        </include>
    </group>

    <!-- DRONE NODES -->
    <group ns="drones">
        <arg name="ns" default="$(arg num_drones)" />
        <group ns="$(eval 'drone' + str(arg('ns')))">
            <node name="uwb_dist_sim_drone" pkg="dist_project" type="uwb_dist_sim_drone.py" />
            <node name="motion_planner_drone" pkg="dist_project" type="motion_planner_drone.py" />
            <node name="kalman_localization_drone" pkg="dist_project" type="kalman_localization_drone.py" />
            <node name="image_processing_drone" pkg="dist_project" type="image_processing_drone.py" />
            <node name="target_estimator_drone" pkg="dist_project" type="target_estimator_drone.py">
                <param name="namespace" value="$(arg ns)" />
            </node>
        </group>
        <include file="$(find dist_project)/launch/start_nodes.launch" if="$(eval arg('ns') - 1 >= 0)">
            <arg name="num_unicycles" value="$(arg num_unicycles)" />
            <arg name="num_drones" value="$(eval arg('ns') - 1)" />
        </include>
    </group>

</launch>